### ISSUE TITLE
Overlap with recent issue raised on the main WASM design repo

### ISSUE THREAD

**Issue by cmuratori** [WebAssembly/multi-memory#20]
I noticed this multi-memory proposal when commenting on this WASM design thread about memory support:

https://github.com/WebAssembly/design/issues/1397

It seems this multi-memory proposal already supports most of what we were asking for in that thread, with the notable exception of decommiting memory to "shrink" memory regions that have previously grown.  Is it possible that memory decommitting could be considered in this proposal, or a separate interrelated proposal?

\- Casey

**Comment by dschuff** [comment]
I think that those are probably separate issues from multiple memories.
I think that shrinking memory, other useful memory management features such as those mentioned in design/#1397, as well as things like mapping files or external memory buffers to reduce copying are all just as valuable when there's only one memory as when there are many. And from a practical implementation and spec point of view, they are quite different. multi-memory requires basically the same VM capabilities that we have now, and no major changes to embedding environments that already support having multiple instances (with their own separate memories) at the same time. 

(OK, now I just went back and read the recent posts on that issue). I think that maybe this proposal isn't as powerful as you think it is. multi-memory allows a module to statically declare some number of memories, and those memories can be independently grown. However the number of memories is fixed, and each load/store instruction has an index which (statically and explicitly) says which memory it refers to. That means it can't easily be used as a substitute for arenas or the various other capabilities of `mmap`-like functionality (because e.g. in C/C++ pointers to those regions must be interchangeable with pointers to any other part of memory, and here they are not).

**Comment by cmuratori** [comment]
I apologize for being clueless about this, but, I wasn't really able to understand this limitation exactly (it was mentioned in the other thread as well).  n memory regions each with 4gb address space is just a 64-bit pointer, with the top 32 bits being the n index, and the bottom 32 bits being the offset in that region.

So, wouldn't a "multi-memory" version of WASM just use 64-bit pointers and then everything "just works"?  The top half of the pointer is the multi-memory index, the bottom half is the offset, and everything proceeds as normal.

The bonus here would be that systems without security concerns can just use direct 64-bit pointers, so there would be no speed penalty on those platforms, either.

Again, apologies for not really knowing the internals of WASM well enough to predict what is wrong with ideas like this one.  I am purely commenting from an external observer's perspective, with the mild panic of knowing I will have to use it as a target but that without prodding it may not support basic memory operations that are fairly important for efficient low-level code :)

\- Casey

**Comment by dschuff** [comment]
Let's be careful about the distinction between the wasm code (i.e. C code compiled to wasm), and the machine code compiled from wasm. Neither the MVP nor multi-memory wasm has 64-bit memory indexes (nor would a memory-shrinking extension, necessarily). A wasm _implementation_ could do what you describe and allocate the memories anywhere in its own 64-bit address space, and then use direct pointers into them. 
But I'm talking about C code compiled to wasm. A C program has a single address space, and any piece of random library code wants to be able to access any memory address. But if there is going to be more than one address space in the wasm output and you have to know at compile time which segment you are addressing, that property won't hold.

**Comment by cmuratori** [comment]
Ah, I see.  I did not realize the MVP was 32-bit only (I assumed the WASM64 idea was going to be part of things).  That is unfortunate, as obviously 32-bit pointer sizes would not make much sense for any kind of scheme.  I was assuming that, for example, a multi-memory proposal could say "pointers are all 64-bit", the C program would therefore be compiled with 64-bit pointers, and then the addressing logic generated for pointer accesses would work as I described.

I guess the appropriate place to look at this will be WASM64, perhaps, then...

\- Casey
